### ISSUE TITLE
>8 bpp in Windows

### ISSUE THREAD

**Issue by jschwartzenberg** [stsp/dosemu2#265]
It would be nice to have support for more than 8 bpp in Windows. There is currently no known working driver which enables this. Some variants:
- Solve the issue that the 15 bpp and 16 bpp Trident TVGA drivers have.
- SciTech Display Doctor (possibly it would have the same issue as the above driver when the crash is fixed.)
- Emulate different video hardware.

Regarding the last point. There are also drivers out there which supposedly would be optimal in combination with a virtual or X11 projected Windows 3.1 environments. These would be the drivers that were used in combination with either OS/2, WABI, Merge/Win4Lin, Desqview/X or SunPCi as those were specificly made to work let the display be a window.

**Comment by jschwartzenberg** [comment]
Multiple versions of the Trident drivers can be found here:
http://www.vogonsdrivers.com/index.php?catid=37&menustate=21,1
In the driver set "Trident TVGA 8800 and 8900 drivers" not all drivers work and at least one doesn't even detect a Trident chip. The 800x600x4 driver from that package does work.

There are also some drivers which have more bpp with a 512x480 resolution, but they also give a corrupted screen.

Not sure if this is relevant, some 1024x768x8+large font drivers trigger this crash in DOSEMU:
```
ERROR: PM stack overflowed: in_dpmi_pm_stack=632
```

**Comment by andrewbird** [comment]
In a former life I used to support a large site with Sun systems, and many of them had SunPCi cards. I don't remember them supporting any Windows versions earlier than 95 and NT, so I just did a scan through some docs over at Oracle(eek!) and I see no mention of earlier Windows installations. Are you sure there are video drivers for Win3.1 / WFW?

**Comment by jschwartzenberg** [comment]
Yep, you're right. I thought maybe the original SunPCi supported Win3.1, but it seems it didn't.

**Comment by andrewbird** [comment]
Another thought is there any such thing as a qxl [driver](https://www.spice-space.org/page/WinQXL) for Win31 so that it can use the spice protocol?

**Comment by andrewbird** [comment]
I seem to remember using DesQview / X too (I might even still have the disks), at the time we were using it to display on remote DEC workstations, and I think the colour depth of those was only 8 bit. But saying that DesQview might have been able to do better. This was certainly in the Win31 era.

**Comment by jschwartzenberg** [comment]
Yeah I was also wondering whether it'd be hard to write a spice plug-in for dosemu. (That would also be a step towards libvirt support.)

**Comment by jschwartzenberg** [comment]
The "SVGA" driver seems to be more stable than the Trident driver. Whenever I play Civilization 2 with the Trident driver, I get this error after my civilization has a new discovery:
![dosemu-win31-trident-driver-crash-after-discovery](https://cloud.githubusercontent.com/assets/877112/20463415/89ef25ce-af32-11e6-8623-f69216306704.png)

The SVGA driver does not have this problem. It doesn't need the VESA patch for dosemu because it supports the exact Trident chip that dosemu emulates already. This info can be found here:
http://sierrahelp.com/Patches-Updates/Patches-Updates-Misc/Win31SVGAUpdate.html

Note that this driver is also limited to 8 bpp.

**Comment by stsp** [comment]
The crash would need a -D9+M log.
Also would be nice if you can make sure it is
not a problem with trident drivers (though I
suspect you can't easily check that).

**Comment by jschwartzenberg** [comment]
Yes,  I indeed did not want to suggest that it's a dosemu problem. It's just an observation. I have no idea if anybody could get access to real hardware. Maybe there are other emulators which also emulate this Trident chip. I'll look around for that when I have time.

**Comment by runderwo** [comment]
I have a TVGA8900 somewhere.

**Comment by stsp** [comment]
> I have a TVGA8900 somewhere.

Then you can see if civ2 works... :)

**Comment by stsp** [comment]
@runderwo in fact, it may be not even all
that difficult. You can just use a "nomodeset=1"
console to pass-through your card. I hope its
a PCI card.

**Comment by stsp** [comment]
And I guess there is no modeset driver for it
in linux anyway, so you don't even need "nomodeset=1"

**Comment by runderwo** [comment]
It's surely an ISA card.  But I could use it to check these drivers for the crash on a real PC with WFW3.1, which I think was the question?

**Comment by stsp** [comment]
If you have a real PC with wfw31 (AFAIK a very rare
wfw - more popular is 3.11), then great.
I just thought no one have such a pc these days, and
pass-through for ISA card may require a bit more
configuration under dosemu, than for pci.

**Comment by stsp** [comment]
It doesn't matter what windows I guess.
What matters is to see if you can get >8bpp
with trident drivers and if civ2 is still stable then.

**Comment by jschwartzenberg** [comment]
I should add that videos are not playing for me in civ2 even though I went through the vfw installer. I can imagine there are still some issues around different aspects of Windows in dosemu which may create such issues.

**Comment by jschwartzenberg** [comment]
I seem that PCem can emulate a TVGA8900D (dosemu has the C variant). I'll see if I can transplant my installation into PCem and see what happens.

**Comment by jschwartzenberg** [comment]
I can confirm that the Trident drivers that work in DOSEMU also work in PCEM. In addition to that, also the 15-bit and 16-bit drivers from that set work correctly in PCEM. Only with 800x600x16, the colors appear a bit strange. The others seems to work without isuse.

So maybe the Trident emulation code from PCEM would be useful. One note, PCEM does require a real Trident ROM to allow its emulation to work.

**Comment by stsp** [comment]
Is the trident driver stable there, or have crash problems?

**Comment by stsp** [comment]
Where do these emulators keep coming from? :)
I am surprised.
How's the speed?

**Comment by jschwartzenberg** [comment]
At the same point in the game, I get this (with the same Trident driver):
![pcem-win31-trident-crash1](https://cloud.githubusercontent.com/assets/877112/21079868/b0b1358a-bf9f-11e6-914e-f43b8e72c822.png)

This is with my dosemu installation moved into PCEM as is. Videos are not playing at all though (neither in dosemu nor in pcem), so maybe there's something messy with my installation due to a dosemu bug, but the normal SVGA driver at least doesn't crash. I would expect the SDD driver to be more stable as well once we get it to work. Otherwise maybe a from scratch installation inside of PCEM and testing with different video chips would reveal which drivers are the more stable ones. I guess fixing up the Trident emulation so that it works with the multitude of drivers out there would be the easiest though.

Maybe there's some tweak I'm missing, but for me the speed of PCEM is very very slow!

**Comment by stsp** [comment]
Thanks for testing!
So why do you suggest fixing the trident
emulation if we see its drivers unstable?
What will this give us other than that "Illegal
instruction" error? OTOH the error message
look very different. Who knows if its the same
problem or not... I don't think you have uploaded
the relevant boot log.

> due to a dosemu bug, but the normal SVGA driver at least doesn't crash.

Do they not crash also under pcem?

> Maybe there's some tweak I'm missing, but for me the speed of PCEM is very very slow!

I looked at the code a bit and it seems it emulates
the hardware much deeply than dosemu or dosbox
do, it is roughly similar to bochs. For example it even
emulates the particular chipsets and can use the native
bioses because of this. In fact, I am very surprised at how
deeply it emulates some HW for which the specs are not
even easily available. It seems to have jit for CPU, but all
other emulators do have it too. So I really am not sure
what market does it target and what is the reason of
competing with bochs, but it should be very handy
for testing things that do not work properly on dosemu.

Could you check if the sound in Mortal Kombat 1 work
properly there? I was never able to get it to work under
dosemu...

**Comment by jschwartzenberg** [comment]
> So why do you suggest fixing the trident emulation if we see its drivers unstable?

1. implementing a completely different chip seems hard
2. there are probably stable drivers, but we need more testing for that (I should try SDD with PCEM)

> Could you check if the sound in Mortal Kombat 1 work properly there? I was never able to get it to work under dosemu...

I guess yeah. I have a script now which copies a dosemu directory into a disk image. With this I can quickly move my dosemu c and d into two separate disk images and assign both as hard disks in PCEM. Other than that I have PCEM configured to use a Trident and a SoundBlaster with the same IRQ as dosemu (rest already matched). Doesn't the sound for this game work in dosbox either? Does it matter which type of SoundBlaster is emulated?

Let me know if I should post this script. The author of PCEM is also very responsive in the forums there btw.

**Comment by stsp** [comment]
> there are probably stable drivers, but we need more testing for that (I should try SDD with PCEM)

Are svga drivers stable under pcem?

> Doesn't the sound for this game work in dosbox either?

It does... but, according to the forum, that somehow
depends on the framerate settings:
http://www.vogons.org/viewtopic.php?p=92472

> Does it matter which type of SoundBlaster is emulated?

The interesting thing is that it does not.
You can select mt32, adlib or even pc-speaker, and it
will still fail to initialize your sound card. That's silly.
But MK2 behaves the same way.

**Comment by jschwartzenberg** [comment]
Are you sure it's not due to your system (dosemu) being too fast? There are quite a few games that fail to detect a SoundBlaster properly when a system is too fast. That's probably one of the reasons that dosbox doesn't run at max speed by default as well.

What happens when you run the game with a slow down utility in dosemu?

**Comment by stsp** [comment]
Not tried...
Maybe you can. :)
But the interesting test would be to try it
on the bare machine, which is fast enough,
but I wonder if the problem is still there.
If so - dosemu is then innocent and we can
try to prepare the binary patch.

**Comment by jschwartzenberg** [comment]
I'll see if I have time for that. The fastest bare machine I have with an ISA slot is 700MHz. I can imagine that some detection code needs a faster system to trip over, so finding a combination of SB support with a fast system will be quite hard. Do modern mainboards still have PCI slot? Maybe it could work with an Audigy + dos drivers then, but that's already partially venturing into emulation imo.

**Comment by stsp** [comment]
Since the problem is reproducible not only with
SB but with any card (even pc-speaker), there should
be no problem to test on _any_ machine.

**Comment by jschwartzenberg** [comment]
I wouldn't vouch for all modern systems to have a proper working pc-speaker though..

**Comment by stsp** [comment]
It doesn't need to be working, it is not detectable. :)
(without looking into pnp tables or smbios, which is what
MK does not do)
Old software thinks it is always there.

**Comment by jschwartzenberg** [comment]
But if you can't hear it or see whether it's detected or not, then how can you test it? I guess I misunderstand what problem is observed. I thought that the audio was not heard and/or the sound chip not detected.

**Comment by stsp** [comment]
No, the program can't detect pc-speaker, so it assumes
it is always there. You, the user, can hear the results and
deduce whether it works correctly or not. So seeing the
message "cannot initialize sound card" even when you
selected pc-speaker in the setup, will mean there is a bug
in the program and not in dosemu.

**Comment by jschwartzenberg** [comment]
I didn't have time to do any tests yet. Browsing around a bit I I found a lot of info about MK1 sound issues. How is MK1 working in dosemu? Does it seem likely they are related to each other?

It appears that at least that game suffers from a detection bug on too fast systems:
https://cloakedthargoid.wordpress.com/mk1-dos-sblaster-fix/
> These bugs are not related to the Adlib / Sound Blaster initialization bug which happens on fast CPUs. That bug is present in both the old and the new versions of the driver. However, it is less of an issue, since it only occurs when you try to configure the audio device. If you bypass it during the configuration (e.g., by temporary slowing down the CPU), it will not happen until you try to change the audio device again.

**Comment by stsp** [comment]
Hmm, you are a good googler. :) I was not able to
find any mentions of that problem. They also mention
the bug-free CD release, but it is not clear whether
this particular bug is fixed there or not (it is sure as
hell not, because it is there even in MK2 - only MK3
got it finally fixed).

**Comment by stsp** [comment]
https://www.vogons.org/viewtopic.php?f=7&t=45709&start=20
Related thread.

**Comment by jschwartzenberg** [comment]
I just tried this version:
https://www.classicdosgames.com/game/Mortal_Kombat_II.html

I can't move the cursor in the F10 menu in dosemu.

**Comment by stsp** [comment]
Confirmed.
Same in mk1.
Perhaps a regression with recent kbd changes.
Please fill in the bug report.

**Comment by stsp** [comment]
> The SVGA driver does not have this problem. It doesn't need the VESA patch for dosemu
> because it supports the exact Trident chip that dosemu emulates already.

Just to clarify: the patch is needed when
you use VGA pass-through in console mode.

**Comment by jschwartzenberg** [comment]
I tried Mortal Kombat II with slowdown from here: http://bretjohnson.us/
When I use it with the argument /MHz486:50 it still can't initialize a Sound Blaster. With slower speeds, the game pretty much runs too slow to get into the menu.

**Comment by stsp** [comment]
What we still dont know for sure is if
this problem manifests on the real hw or
not. I guess the bootable flash will help
finding this out...

**Comment by stsp** [comment]
> Not sure if this is relevant, some 1024x768x8+large font drivers trigger this crash in DOSEMU:

For me it just exits, without any crash.
There is a GPF in the log though, and on a winos2
kernel too, so I don't suppose we'll be able to get
it to work. May be worth a try on another emu.
Btw, you should try the civ problem with the winos2
kernel to see if its there or not.

**Comment by jschwartzenberg** [comment]
Yeah, the winos2 kernel sounds like a good idea. I was searching around a bit and ran into this:
http://tech-insider.org/personal-computers/research/1993/0308.html
Maybe the Trident drivers are picky in general. I was also thinking of trying to configure the SDD drivers inside PCEM and then copying the configured SDD into dosemu.

**Comment by jschwartzenberg** [comment]
> What we still dont know for sure is if this problem manifests on the real hw or not.

I tried the MK2 demo on real hardware yesterday. Pentium 3 700MHz with an Audigy2 (one of the last cards that still has DOS drivers I guess) and with that setup I can select the Sound Blaster option and the sound works.

**Comment by stsp** [comment]
Sigh.
Anyway, I think this should go to another ticket.

**Comment by jschwartzenberg** [comment]
A small update. A driver that does 16 bpp at 512x480 is available here:
http://vetusware.com/download/TRIDENT%20TVGA%20Drivers%20For%20Windows%203.1%20UD5.5/?id=10229

For some reason, this driver is commented out by default in oemsetup.inf, so you have to remove the ; to make it visible for Windows Setup.

![dosemu-win31-16bpp](https://user-images.githubusercontent.com/877112/27002928-48f32890-4ded-11e7-86ad-61f631481a84.png)

**Comment by jschwartzenberg** [comment]
This driver can be found in multiple other places with the name ud55. Searching online for ud55-win.exe, ud55-win.zip, ud55.exe, ud55.zip should help to find it in other places as well.

Similarly a set with uc51 exists. The driver sets named u?5? are the last drivers sets that install properly and from which multiple drivers work correctly. For any set named u?6? the installer will fail to detect a Trident chip.
![dosemu-win31-us60](https://user-images.githubusercontent.com/877112/27003520-8c7d02ec-4df8-11e7-9ede-c4a076674afd.png)

uc51 and ud55 can be installed next to each other. The included setup tool will show both driver sets installed and allow picking a driver:
![dosemu-win31-trident](https://user-images.githubusercontent.com/877112/27003522-9bf4660c-4df8-11e7-8529-029728448227.png)

There is also a set which would be uc52 according to its readme.txt but I could only find it only as Tri8900.zip. This set has a date of May 1994 (uc51 and ud55 are from October 1993). Its setup utility is different. Some of its drivers don't work well with dosemu and some are binary identical to the ones from uc51.
![dosemu-win31-trident2](https://user-images.githubusercontent.com/877112/27003581-d640dd1c-4df9-11e7-93f7-8a4ea86d4af1.png)

**Comment by stsp** [comment]
But does civ2 work?
Would be good to update the docs with
these findings!

**Comment by jschwartzenberg** [comment]
Yes, it runs, even with the videos in `win /3` mode. I still have to look at the music though, I think it wants to play CD audio for that.

**Comment by stsp** [comment]
Yes but it was not stable enough.

**Comment by jschwartzenberg** [comment]
In that set-up, I managed to play it for hours. I forgot which driver I used then, but it should be easy to figure that out :)

**Comment by stsp** [comment]
So seems there is a bug in many 8bpp drivers
that breaks civ2, and not a dosemu problem?

**Comment by jschwartzenberg** [comment]
I think the bug was only triggered with the Trident driver with Windows in 286 mode. The Indeo videos from Civ2 require 386 mode. Probably it won't happen in 386 mode at all. This is all quite a while ago though. Let me know if you'd be interesting in trying the game itself. I use CDemu to mount the image. I'll probably have to use the CDROM.SYS driver on the device instead of `-cdrom` to be able to get the music to work.

**Comment by jschwartzenberg** [comment]
It seems this tool runs in DOSEMU: https://winworldpc.com/download/65c2b8e2-8094-c3b8-11c3-a311c3a7cb86
Not really useful though :)

**Comment by stsp** [comment]
Add a screen-shot if its something interesting.

**Comment by jschwartzenberg** [comment]
https://www.youtube.com/watch?v=sOYcY6oeBhY

**Comment by stsp** [comment]
What is winx?

**Comment by jschwartzenberg** [comment]
Here is more information about it: http://toastytech.com/guis/dvx3.html

I suspect getting DESQview/X to run to be tricky though.

**Comment by stsp** [comment]
So an rdesktop for win31, cool!

**Comment by jschwartzenberg** [comment]
Not sure how you see that. You mean RDP server for Windows 3.1? It transforms Windows 3.1 into an X11 application.

**Comment by stsp** [comment]
> DESQview/X includes a program that overrides the Windows 3.1 video, keyboard,
> and sound drivers to turn Windows 3.1 in to an X-window application!

This is basically what rdesktop does:
it also passes the input and output
via network, having an X application
(for drawing and grabbing input) at
a local end.

> You mean RDP server for Windows 3.1?

Just not with RDP protocol. :)

**Comment by jschwartzenberg** [comment]
I think rdesktop is just a client though, but yeah it gets the point :)

**Comment by andrewbird** [comment]
Is it possible to write a Win3.1 display driver that writes to the SDL window directly?

**Comment by stsp** [comment]
What kind of SDL under DOS?

**Comment by andrewbird** [comment]
I meant a Win31 display driver that instead of drawing in dosemu's virtual vga card it calls out to dosemu 32 bit code to draw in linux SDL window. I'm guessing most of the Windows display driver interface would be high level functions (draw line, clear screen, write text) and that perhaps the meat of any implementation could be done in dosemu. I've previously considered doing something similar for Borland's .BGI interface, but I could never find any commercial DOS apps that used .BGI, so it didn't seem worth the effort.

**Comment by stsp** [comment]
I believe this is mostly what winx does.
But for dosemu there is no big advantage
of para-virtualized video over the vesa lfb.

**Comment by andrewbird** [comment]
Perhaps large display size and high colour depth? Anyway I'm sure it's a lot of work.

**Comment by stsp** [comment]
It doesn't give you anything over LFB,
really. You will still need to map SDL
drawing surface to DOS mem and pass
the pointer to the driver. Same as LFB,
no difference at all. The differences may
start only when the OS doesn't want to
use v86 for vesa calls, which is the case
with linux (unless you install uvesafb+vm86d,
which is a gross hack). But for windows-10
or 3.1 there is no problem to get an LFB pointer
via int10 VBE calls.

**Comment by stsp** [comment]
> Anyway I'm sure it's a lot of work.

Its not, its the same thing.

**Comment by stuaxo** [comment]
The one advantage a custom display driver could have is it would be possible to make it able to change the virtual screen size when the window is resized.

**Comment by stsp** [comment]
Resized in what way?
With mouse drag?

**Comment by stuaxo** [comment]
Yes, by mouse drag

**Comment by stsp** [comment]
And what is "virtual screen size"?
Resolution of win31?
I really really doubt such thing is
possible in win31. IIRC you need
to reboot it to switch resolution.
Maybe it is somehow possible in
windows-10, eg the driver sends
some system event, and all scalers,
font renderers and alike - adjust
their settings.

**Comment by stuaxo** [comment]
The amount of pixels in X and Y is changed, but other VM hosts don't seem to worry about scalers etc.

Windows XP and up in VM hosts such as vmware/virtuabox/qemu can do this when the user scales the window on the host:
https://www.youtube.com/watch?v=wPvDCK-bbAI

The user scales the window and the VM host + guest agent handle the device resize, and telling windows to redraw everything.

To avoid having to juggle where everything is drawn to, every resize, exploiting VESAs pitch/bytes per horizontal line may be an option.

VESA allows each line to have more bytes per line than needed to draw pixels for the screen size.

A custom mode could be created by DOSEMU that is big enough to draw on the entire host screen, but with the bytes per horizontal line setup to draw only enough pixels needed to fill the host Window.

On resize, these values would be fixed up and the custom driver would then tell Win 3.1 to redraw each window.

In modern windows there is a RedrawWindow API https://docs.microsoft.com/en-gb/windows/win32/api/winuser/nf-winuser-redrawwindow?redirectedfrom=MSDN

This probably doesn't exist in Windows 3.1, but maybe sending WM_PAINT to each Window would suffice.

**Comment by stsp** [comment]
> The amount of pixels in X and Y is changed, but other VM hosts don't seem
> to worry about scalers etc.

Yes, I was talking about a VM guest, not host.

> Windows XP and up in VM hosts such as vmware/virtuabox/qemu can do this

But win31 is not in the list of "XP and up".
It requires reboot after resolution change,
and as such, can't do this trick.

> On resize, these values would be fixed up and the custom driver would then tell
> Win 3.1 to redraw each window.

Its not possible to say to win31 "we have
a virtual resolution of 1920x1080, but please
use only 640x480 until we tell you to re-draw
in another size". Driver can't tell win31 what
resolution currently is, with WM_PAINT event
or by any other means.

**Comment by jschwartzenberg** [comment]
I imagine that an X11fied win31 might not work at the driver level, but at higher level. Individual windows themselves already support resizing. Possibly the desktop "behind" them can be flexible too. It would be interesting to try out winx.

Note that the drivers from Merge/Win4Lin also have X11 in their name. Possibly those drivers use drawing methods that align with X11. They are display drivers though, so you're stuck with a fixed resolution as well.

**Comment by stsp** [comment]
Doesn't winx require desqview?
Also I am not sure why do you think
that way, as the description of winx
seem to say its just a driver.
I think you would then need to
re-implement entire GDI and explorer.exe.

**Comment by jschwartzenberg** [comment]
I guess it does. I suppose it depends at least on some desqview specific hook if winx has some X11ness in it. It would be nice to run win31 in desqview in dosemu2 and export the win31 window via dosemu's network support to the host display :)

**Comment by stuaxo** [comment]
Looks like it may well be able to export the windows over the network with X, according to this:
http://www.cs.cmu.edu/~bmm/dvx.html

>  It allows you to run Windows 3.1 (but not Windows 95/98, NT, or 2000) in an X-window, and display on any X-server. This means that while sitting at your Unix workstation you can run Windows programs remotely ...

**Comment by stuaxo** [comment]
Here's an old guide for using DV/X apps on a real DOS computer from X on a networked Linux computer
https://www.linuxjournal.com/article/5542

**Comment by jschwartzenberg** [comment]
Maybe it's possible to get this running in PCEM and export the display to the host. I'm not sure whether QEMU would handle it.

**Comment by stuaxo** [comment]
Virtuallyfun covered it in qemu and exporting to X
https://virtuallyfun.com/wordpress/2011/03/27/desqviewx/

It seems to want the QEMM to run which could be an obstacle here (though who knows if it checks for QEMM or some extension it needs)

**Comment by jschwartzenberg** [comment]
I guess it might work with Windows too then. I wasn't sure because QEMU is known to be more flacky with Windows 9x, especially with KVM. Maybe QEMM would actually help to get Windows 3.1/9x more stable in QEMU :)

**Comment by stsp** [comment]
I wonder if qemm had a significant use
in win9x (it had a use in win31 only by the
means of desqview, and even that didnt
seem to work in 9x). In the first IT lecture
in university 25 years ago, the
professor asked us to instantly forget
all those many books about XT-style
memory management and its 386
extensions, as win95 makes this whole
industry irrelevant.

**Comment by jschwartzenberg** [comment]
[msdos622+win31+dvx.qcow2.gz](https://github.com/dosemu2/dosemu2/files/4638778/msdos622%2Bwin31%2Bdvx.qcow2.gz)
Runs with: `qemu-system-x86_64 -hda ./msdos622+win31+dvx.qcow2` inside QEMU type `dvx`

**Comment by stsp** [comment]
Looks absolutely impressive!
Until you run win31 directly (not winx),
in which case I got mouse stuck.
Why didn't quarterdeck market their
own OS then? :)

**Comment by jschwartzenberg** [comment]
No idea, I also can't move or resize any windows in DVX. I have the feeling something is not working as intended. Maybe due to QEMU or CuteMouse. Without CuteMouse, the mouse doesn't work at all in DVX, but maybe another driver works better.

**Comment by jschwartzenberg** [comment]
I tried the latest versions of the Microsoft & Logitech mouse drivers from https://www.uwe-sieber.de/util_e.html, but with those loaded the mouse doesn't work in DVX.

**Comment by stsp** [comment]
Check them in dosemu2 with win31 and
my win31 driver first.

**Comment by stsp** [comment]
Note that the driver was recently updated.

**Comment by jschwartzenberg** [comment]
Naively running DVX in DOSEMU2:
![image](https://user-images.githubusercontent.com/877112/82131277-c78cab80-97d3-11ea-869c-261fbc5bc3f7.png)

Got a crash with that option, but I'm not fully up to date: [boot.log](https://github.com/dosemu2/dosemu2/files/4639316/boot.log)

Note that I just did this:
```
qemu-img convert -O raw msdos622+win31+dvx.qcow2 msdos622+win31+dvx.img
env dosemu__hdimage=/home/julius/desqviewx/msdos622+win31+dvx.img dosemu
```

**Comment by stsp** [comment]
I meant you should check the DOS mouse
drivers with my int33 mouse driver for win31,
just to see if they work properly. As you seem
to have problems finding a functional one.
I didn't say you can run desqview.

**Comment by stsp** [comment]
I.e. not all mouse drivers are compatible
with "confined" win31, but you can check
them with dosemu2 first.

**Comment by jschwartzenberg** [comment]
As I suspected, the problem are due to QEMU. I converted to image to raw and opened it with PCEM. This runs better. Maybe tweaking the hardware emulation settings would help QEMU.

It's not possible to make the MS Windows window bigger, it can only be made smaller and that will give it scroll bars:
![image](https://user-images.githubusercontent.com/877112/82733038-6bfd7900-9d11-11ea-8167-26233ce38d94.png)

Now I need to see if it can be exported to my host X.org X11 server.

**Comment by jschwartzenberg** [comment]
Somebody installed WABI in a VM and is providing an image to download:
https://virtuallyfun.com/wordpress/2020/11/11/fun-with-caldera-wabi/

I haven't checked it out yet.

Someone is also linking to this document, but it is not clear if this also applies to the Caldera x86 version:
https://www.hotchips.org/wp-content/uploads/hc_archives/hc08/2_Mon/HC8.S2/HC8.2.1.pdf

**Comment by stsp** [comment]
Yes, quite an interesting reading.
Seems like an early wine: implements
USER.DLL, GDI.DLL, KERNEL.DLL
and the rest should be taken from windows.
Wine did the same in 90x IIRC.

No sources of course?

**Comment by jschwartzenberg** [comment]
> No sources of course?

I don't think so. I guess they had access to the Windows 3.1 sources. It'd be interesting to check out if they indeed ship their own parts of Windows 3.1.

**Comment by stsp** [comment]
I think they do, but it doesn't mean you'll
see them in an accordingly named dll files
there. :)

**Comment by jschwartzenberg** [comment]
I'm curious whether win32s works with WABI :)

**Comment by stsp** [comment]
Likely not at all.
But even if it does, it would be via
the custom win32s dll rather than
the vxd driver.
In dosemu we do not re-implement
dlls.

**Comment by stsp** [comment]
So this is solved by vgesvga.
